core/Archive/DataTableFactory.php

```diff
@@ -157,21 +161,62 @@ public function useSubtable($idSubtable)
      */
     public function make($index, $resultIndices)
     {
+        $keyMetadata = $this->getDefaultMetadata();
+
         if (empty($resultIndices)) {
             // for numeric data, if there's no index (and thus only 1 site & period in the query),
             // we want to display every queried metric name
             if (empty($index)
-                && $this->dataType == 'numeric'
+                && $this->isNumericDataType()
             ) {
                 $index = $this->defaultRow;
             }
 
-            $dataTable = $this->createDataTable($index, $keyMetadata = array());
+            $dataTable = $this->createDataTable($index, $keyMetadata);
         } else {
-            $dataTable = $this->createDataTableMapFromIndex($index, $resultIndices, $keyMetadata = array());
+            $dataTable = $this->createDataTableMapFromIndex($index, $resultIndices, $keyMetadata);
         }
 
-        $this->transformMetadata($dataTable);
+        return $dataTable;
+    }
+
+    /**
+     * Creates a merged DataTable|Map instance using an index of archive data similar to {@link make()}.
+     *
+     * Whereas {@link make()} creates a Map for each result index (period and|or site), this will only create a Map
+     * for a period result index and move all site related indices into one dataTable. This is the same as doing
+     * `$dataTableFactory->make()->mergeChildren()` just much faster. It is mainly useful for reports across many sites
+     * eg `MultiSites.getAll`. Was done as part of https://github.com/piwik/piwik/issues/6809
+     *
+     * @param array $index @see DataCollection
+     * @param array $resultIndices an array mapping metadata names with pretty metadata labels.
+     *
+     * @return DataTable|DataTable\Map
+     * @throws \Exception
+     */
+    public function makeMerged($index, $resultIndices)
+    {
```

Comment:
I think instead of makeMerged + makeMergedTableWithPeriodAndSiteIndex/makeMergedWithSiteIndex, this method could be made generic w/ a `makeMerged($index, $resultIndices, $mergeIndices)`. I haven't coded an implementation so I don't know if there are any hidden problems, but I think it work by:

1. creating a nested DataTable\Map using values in $resultIndices (unless it's empty)
2. then creating individual tables & setting columns using the keys in $mergeIndices

Generic code means cleaner, reasoned code, and thus less technical debt. Hopefully, it's possible w/o optimization trade-offs.